src/main/scala/nl/knaw/dans/easy/multideposit/MultiDepositParser.scala

```diff
@@ -17,79 +17,77 @@ package nl.knaw.dans.easy.multideposit
-        val output = parser.getRecords.filter(r => r.size() > 0 && !r.get(0).isBlank).map(_.toList)
-        validateDatasetHeaders(output.head)
-          .map(_ => {
-            val csvData = output.tail.map(output.head zip _)
-            log.debug("Successfully loaded CSV file")
-            Observable.from(csvData)
-              .zipWithIndex
-              .map { case (xs, i) => (xs, i + 2) } // +2 because we want the second line (element 0) to have index 2
-              .foldLeft(new Datasets)((dss, tuple) => updateDatasets(dss, tuple))
-          })
-          .onError(Observable.error)
-          .subscribe(subscriber)
-      }.onError(subscriber.onError)
-    })
+        for {
+          record <- parser.getRecords.asScala
+          if record.size() > 0
+          if !record.get(0).isBlank
+        } yield record.asScala.toList
+      }
+      _ <- validateDatasetHeaders(content.head)
+    } yield {
+      debug("Successfully loaded CSV file")
+      content.tail
+        .map(content.head.zip(_))
+        .zipWithIndex
+        .foldRight(Datasets) {
+          // +2 because we want the second line (element 0) to have index 2
+          case ((values: CsvValues, row: Int), dss: Datasets) => updateDatasets(dss, values, row + 1)
+        }
+    }
   }
 
   /**
-    * Tests whether the given list of headers is valid. If so, `Success(Unit)` is returned,
-    * else a `Failure` with a detailed `ActionException` is returned.
-    *
-    * @param headers the headers to be validated
-    * @return `Success` if the `headers` are valid, `Failure` if the `headers` are invalid
-    */
+   * Tests whether the given list of headers is valid. If so, `Success(Unit)` is returned,
+   * else a `Failure` with a detailed `ActionException` is returned.
+   *
+   * @param headers the headers to be validated
+   * @return `Success` if the `headers` are valid, `Failure` if the `headers` are invalid
+   */
+  // TODO valid headers listing as sub-command???
   def validateDatasetHeaders(headers: List[String]): Try[Unit] = {
     val fileHeaders = List("FILE_SIP", "FILE_DATASET", "FILE_AUDIO_VIDEO", "FILE_STORAGE_SERVICE", "FILE_STORAGE_PATH", "FILE_SUBTITLES")
     val springfieldHeaders = List("SF_DOMAIN", "SF_USER", "SF_COLLECTION", "SF_PRESENTATION", "SF_SUBTITLES")
     val administrativeHeaders = List("DEPOSITOR_ID")
 
     val validHeaders = "DATASET" :: fileHeaders ++ springfieldHeaders ++ administrativeHeaders ++ DDM.allFields
-    if (headers.forall(validHeaders.contains)) Success(Unit)
-    else Failure(new ActionException(0, "SIP Instructions file contains unknown headers: "
-      + headers.filter(!validHeaders.contains(_)).mkString(", ") + ". "
+    val invalidHeaders = headers.filterNot(validHeaders.contains)
+    if (invalidHeaders.isEmpty) Success(Unit)
+    else Failure(ActionException(0, s"SIP Instructions file contains unknown headers: ${invalidHeaders.mkString(", ")}. "
```

Comment:
We have inconsistent behaviour between components, [one](https://github.com/DANS-KNAW/easy-update-metadata-dataset/blob/8e95874/src/main/scala/nl/knaw/dans/easy/umd/InputRecord.scala#L27-L50) requiring a specific order, the other not. Which one should change?